### ISSUE TITLE
NativeAOT publish for DLL does not set automatically RID

### ISSUE THREAD

**Issue by kant2002** [dotnet/runtime#80338]
I was have following setup. DLL project which I publish as NativeAOT. I have to publish it `dotnet publish -r win-x64 --self-contained` otherwise project does not pickup nor current RID of my dev machine, not the fact that application is self-contained (NativeAOT). It's most likely issue in SDK

- `dotnet publish` want me specify RID
- `dotnet publish -r win-x64` use WinForms references during build, but do not propagate them for publish to ILC
- `dotnet publish -r win-x64 --self-contained` works, but `self-contained` switch is misleading probably and undo a lot of tooling work in recent releases around NativeAOT

**Comment by am11** [comment]
In .NET 7+, dotnet-publish sets the RID implicitly to the runtime of executing dotnet.exe with `Publish{Aot,SingleFile,ReadyToRun}`.

You can also pass `--use-current-runtime` or `--ucr` in a general-purpose script (e.g. for CI) for cases which aren't covered by implicit RID. I don't remember offhand which are those cases, @nagilson might.

`dotnet new mvc -n mvc1 && dotnet publish -p:PublishAot=true mvc1` publishes AOT app for win-x64 on Windows 10 x64 machine; without explicitly specifying the runtime with .NET 7 (and 8 alpha).

**Comment by kant2002** [comment]
I know about that, but if you create 
```bash
dotnet new winforms 
# remove OutputType
dotnet publish
```
then you will have situation like I describe. Probably MVC is special cased somehow.

**Comment by am11** [comment]
Well, I tested with two templates: Console and MVC. WinForms is rather an exception.

> ```shell
> dotnet new winforms 
> # remove OutputType
> dotnet publish
> ```

I don't see PublishAot specified anywhere.

**Comment by kant2002** [comment]
You may take a look at this project https://github.com/kant2002/WinFormsComInterop/blob/main/samples/HostedWindowsForms/HostedWindowsForms.csproj

which was led to following error https://github.com/kant2002/WinFormsComInterop/issues/48#issue-1504667504 if using `dotnet publish -r win-64`.

**Comment by kant2002** [comment]
I think issue is specifically for `OutpuType=Library` kind of projects. That maybe by design in regular CoreCLR publish, but unexpected for NativeAOT

**Comment by am11** [comment]
Yes, classlib requires the runtime argument:

```sh
$ dotnet new classlib -n lib1
$ dotnet publish lib1 -p:PublishAot=true --use-current-runtime
```

**Comment by kant2002** [comment]
I see two problems here, why `PublishAot` does not imply `--use-current-runtime` and second. Why `PublishAot` does not imply `--self-contained`. I may probably understand some rationale for first case, but not for second.

**Comment by am11** [comment]
Your sample https://github.com/kant2002/WinFormsComInterop/tree/main/samples/HostedWindowsForms is building on my Win10 box with .NET SDK 7.0.101 without `--self-contained`:

```cmd
dotnet publish --use-current-runtime
::: or
dotnet publish -r win-x64

...
Generating native code
...
```
produces `bin\Debug\net7.0-windows\win-x64\publish\HostedWindowsForms.dll` which is a native library according to ildasm.exe.

**Comment by kant2002** [comment]
I belive you are did not notice `ILC: Method '[HostedWindowsForms]HostedWindowsForms.LibraryFunctions.ShowMessageBox()' will always throw because: Fai
  led to load assembly 'System.Windows.Forms'` . that's because System.Windows.Forms.dll is not part of set of published files. And that's because dotnet SDK think this is not self-contained app.

**Comment by am11** [comment]
> `ILC: Method '[HostedWindowsForms]HostedWindowsForms.LibraryFunctions.ShowMessageBox()' will always throw because: Fai led to load assembly 'System.Windows.Forms'`

Sounds like this should be an error.

**Comment by kant2002** [comment]
ILC reporting is separate issue. I'm worry only about SDK behavior. It seems reasonable that if `PublishAot=true` then SDK infer RID and `--self-contained` independently from OutputType of the project.

**Comment by MichalStrehovsky** [comment]
This is probably because publishing a library mostly works by accident in the SDK and wasn't fully tested for everything. https://github.com/dotnet/linker/issues/3165#issuecomment-1369979182. Looks like it doesn't even work for all NativeAOT scenarios (although WinForms with NativeAOT is definitely not a supported scenario in 7.0).

> > `ILC: Method '[HostedWindowsForms]HostedWindowsForms.LibraryFunctions.ShowMessageBox()' will always throw because: Fai led to load assembly 'System.Windows.Forms'`
> 
> Sounds like this should be an error.

Nope, people add assemblies their grandmother wrote for .NET Framework 1.0 to their .NET 7 apps and expect the parts that work with a JIT-based runtime to also work on an AOT-based runtime. (I.e. publishing apps with missing or mismatched dependencies is common so the compiler makes a reasonable attempt to emulate the JIT behavior where one would get an exception only once a problematic codepath is reached.)

**Comment by kant2002** [comment]
So we have one more case in addition to Android where publish self-contained libraries make sense. Is this should be better documented in sample, or SDK work?

**Comment by MichalStrehovsky** [comment]
It already mostly works for NativeAOT. Did you hit it in a scenario that doesn't involve WinForms? We don't document/make samples for things that are unsupported.

(I assume the problem is that WinForms adds itself to publish differently than regular NuGet packages and we need a fix in the .NET SDK or WinForms SDK to do it in a way that works with NativeAOT. Basically the SDKs don't expect publishing as libraries and may need extra work.)

**Comment by kant2002** [comment]
Does failure of embed of ASP.NET Core count as "supported" scenario? I test it and it produce `ILC: Method '[webembedded]LibraryFunctions.ShowMessageBox()' will always throw because: Failed to load assembly 'Microsoft.AspNetCore'`. Do you need repo with repro?

**Comment by MichalStrehovsky** [comment]
ASP.NET Core is also not supported with NativeAOT. It's only a .NET 7 servicing-worthy issue if this doesn't work with regular NuGets. Otherwise this would be "help-wanted", milestoned as "Future".

**Comment by kant2002** [comment]
I think it affects only cases where library distinguish between self-contained and not self-contained modes. I do not aware of non-MS and popular enough Sdks which do this. So be it -"help wanted" and "Future". Any chance that sdk repo(am I right?) would accept contributions in this area? Once ASP.NET Core make significant progress in NativeAOT I think this issue will re-appear, since I bet some people try to put it in the unmanaged app.

**Comment by am11** [comment]
If most (or all) of the `PublishAot=true` cases must imply `SelfContained=true` ignoring the project `OutputType`, then this should be fine: https://github.com/am11/sdk/commit/dc23244c6a5fd0bddbd24ebab895402be1bc8f0f (props are still overridable if someone really needs SelfContained=false).

**Comment by vitek-karas** [comment]
NativeAOT has the unwanted privilege of being basically the first scenario in which library project needs to be published self-contained. (publishing library project itself is somewhat undefined operation in SDK, making it self-contained is just making it worse). Android apps are like this already, but I think they go well outside the normal SDK integration, so it's not a fair comparison.

> Any chance that sdk repo(am I right?) would accept contributions in this area? 

I don't see why it wouldn't. I think the toughest part is actually defining what it means to do these things. And then reconciling that with existing behavior which is most likely different and thus it will be a breaking change.

**Comment by nagilson** [comment]
This is definitely an SDK issue, but I don't agree with the idea of turning the automatic RID on for everything... in fact it was like that during part of 7.0.100, and it broke a lot of people.

But @kant2002 is 💯 % right in that it's the fact that it's a library that's the reason there is no implicit RID. The RID only is inferred for executable projects, there are a lot of library projects: EX for a classlib, `PublishReadyToRun` should not build or publish successfully.  

Maybe the suggestion is for `winforms` project types in particular we allow them to get the implicit RID? Tbh I don't know anything about winforms, but it sounds like something we'd want to do.
